### ISSUE TITLE
NoSuchElementException on concurrent requests to RegistryClientFactory.create

### ISSUE THREAD

**Issue by MosheElisha** [Apicurio/apicurio-registry#2132]
Hi,

I have JUnit tests running in parallel and these tests are using `org.springframework.kafka.annotation.KafkaListener` annotation.
Sometimes, I get the exception below. 

I might be wrong but from what I understand, the method `io.apicurio.registry.rest.client.RegistryClientFactory#resolveProviderInstance` is not thread-safe and concurrent requests to `io.apicurio.registry.rest.client.RegistryClientFactory#create(java.lang.String, java.util.Map<java.lang.String,java.lang.Object>, io.apicurio.rest.client.auth.Auth)` on init might cause `NoSuchElementException`.

```
org.apache.kafka.common.KafkaException: Failed to construct kafka consumer
	at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:819) ~[kafka-clients-2.7.0.jar:?]
	at org.springframework.kafka.core.DefaultKafkaConsumerFactory.createRawConsumer(DefaultKafkaConsumerFactory.java:366) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.springframework.kafka.core.DefaultKafkaConsumerFactory.createKafkaConsumer(DefaultKafkaConsumerFactory.java:334) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.springframework.kafka.core.DefaultKafkaConsumerFactory.createConsumerWithAdjustedProperties(DefaultKafkaConsumerFactory.java:310) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.springframework.kafka.core.DefaultKafkaConsumerFactory.createKafkaConsumer(DefaultKafkaConsumerFactory.java:277) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.springframework.kafka.core.DefaultKafkaConsumerFactory.createConsumer(DefaultKafkaConsumerFactory.java:254) ~[spring-kafka-2.7.9.jar:2.7.9]
	at io.opentracing.contrib.kafka.spring.TracingConsumerFactory.createConsumer(TracingConsumerFactory.java:97) ~[opentracing-kafka-spring-0.1.15.jar:?]
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.<init>(KafkaMessageListenerContainer.java:717) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.springframework.kafka.listener.KafkaMessageListenerContainer.doStart(KafkaMessageListenerContainer.java:320) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.springframework.kafka.listener.AbstractMessageListenerContainer.start(AbstractMessageListenerContainer.java:397) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.springframework.kafka.listener.ConcurrentMessageListenerContainer.doStart(ConcurrentMessageListenerContainer.java:205) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.springframework.kafka.listener.AbstractMessageListenerContainer.start(AbstractMessageListenerContainer.java:397) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.springframework.kafka.config.KafkaListenerEndpointRegistry.startIfNecessary(KafkaListenerEndpointRegistry.java:327) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.springframework.kafka.config.KafkaListenerEndpointRegistry.registerListenerContainer(KafkaListenerEndpointRegistry.java:203) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.springframework.kafka.config.KafkaListenerEndpointRegistrar.registerEndpoint(KafkaListenerEndpointRegistrar.java:238) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.springframework.kafka.annotation.KafkaListenerAnnotationBeanPostProcessor.processListener(KafkaListenerAnnotationBeanPostProcessor.java:565) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.springframework.kafka.annotation.KafkaListenerAnnotationBeanPostProcessor.processKafkaListener(KafkaListenerAnnotationBeanPostProcessor.java:459) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.springframework.kafka.annotation.KafkaListenerAnnotationBeanPostProcessor.postProcessAfterInitialization(KafkaListenerAnnotationBeanPostProcessor.java:361) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsAfterInitialization(AbstractAutowireCapableBeanFactory.java:455) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1808) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:431) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.injectDependencies(DependencyInjectionTestExecutionListener.java:120) ~[spring-test-5.3.13.jar:5.3.13]
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.prepareTestInstance(DependencyInjectionTestExecutionListener.java:83) ~[spring-test-5.3.13.jar:5.3.13]
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:248) ~[spring-test-5.3.13.jar:5.3.13]
	at org.springframework.test.context.junit.jupiter.SpringExtension.postProcessTestInstance(SpringExtension.java:138) ~[spring-test-5.3.13.jar:5.3.13]
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeTestInstancePostProcessors$8(ClassBasedTestDescriptor.java:363) ~[junit-jupiter-engine-5.8.1.jar:5.8.1]
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.executeAndMaskThrowable(ClassBasedTestDescriptor.java:368) ~[junit-jupiter-engine-5.8.1.jar:5.8.1]
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeTestInstancePostProcessors$9(ClassBasedTestDescriptor.java:363) ~[junit-jupiter-engine-5.8.1.jar:5.8.1]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177) ~[?:?]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
	at java.util.stream.StreamSpliterators$WrappingSpliterator.forEachRemaining(StreamSpliterators.java:312) ~[?:?]
	at java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:735) ~[?:?]
	at java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734) ~[?:?]
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658) ~[?:?]
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeTestInstancePostProcessors(ClassBasedTestDescriptor.java:362) ~[junit-jupiter-engine-5.8.1.jar:5.8.1]
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$instantiateAndPostProcessTestInstance$6(ClassBasedTestDescriptor.java:283) ~[junit-jupiter-engine-5.8.1.jar:5.8.1]
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.instantiateAndPostProcessTestInstance(ClassBasedTestDescriptor.java:282) ~[junit-jupiter-engine-5.8.1.jar:5.8.1]
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$testInstancesProvider$4(ClassBasedTestDescriptor.java:272) ~[junit-jupiter-engine-5.8.1.jar:5.8.1]
	at java.util.Optional.orElseGet(Optional.java:369) ~[?:?]
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$testInstancesProvider$5(ClassBasedTestDescriptor.java:271) ~[junit-jupiter-engine-5.8.1.jar:5.8.1]
	at org.junit.jupiter.engine.execution.TestInstancesProvider.getTestInstances(TestInstancesProvider.java:31) ~[junit-jupiter-engine-5.8.1.jar:5.8.1]
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$prepare$0(TestMethodTestDescriptor.java:102) ~[junit-jupiter-engine-5.8.1.jar:5.8.1]
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:101) ~[junit-jupiter-engine-5.8.1.jar:5.8.1]
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:66) ~[junit-jupiter-engine-5.8.1.jar:5.8.1]
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$prepare$2(NodeTestTask.java:123) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.prepare(NodeTestTask.java:123) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:90) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService$ExclusiveTask.compute(ForkJoinPoolHierarchicalTestExecutorService.java:185) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:189) ~[?:?]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290) ~[?:?]
	at java.util.concurrent.ForkJoinTask.doJoin(ForkJoinTask.java:396) ~[?:?]
	at java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:721) ~[?:?]
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService.joinConcurrentTasksInReverseOrderToEnableWorkStealing(ForkJoinPoolHierarchicalTestExecutorService.java:162) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService.invokeAll(ForkJoinPoolHierarchicalTestExecutorService.java:136) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService$ExclusiveTask.compute(ForkJoinPoolHierarchicalTestExecutorService.java:185) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:189) ~[?:?]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290) ~[?:?]
	at java.util.concurrent.ForkJoinTask.doJoin(ForkJoinTask.java:396) ~[?:?]
	at java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:721) ~[?:?]
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService.joinConcurrentTasksInReverseOrderToEnableWorkStealing(ForkJoinPoolHierarchicalTestExecutorService.java:162) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService.invokeAll(ForkJoinPoolHierarchicalTestExecutorService.java:136) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139) ~[junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73) [junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138) [junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95) [junit-platform-engine-1.8.1.jar:1.8.1]
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService$ExclusiveTask.compute(ForkJoinPoolHierarchicalTestExecutorService.java:185) [junit-platform-engine-1.8.1.jar:1.8.1]
	at java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:189) [?:?]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290) [?:?]
	at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020) [?:?]
	at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656) [?:?]
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594) [?:?]
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183) [?:?]
Caused by: java.lang.IllegalStateException: java.util.NoSuchElementException
	at io.apicurio.registry.serde.AbstractSchemaResolver.configure(AbstractSchemaResolver.java:86) ~[apicurio-registry-serde-common-2.1.3.Final.jar:?]
	at io.apicurio.registry.serde.DefaultSchemaResolver.configure(DefaultSchemaResolver.java:56) ~[apicurio-registry-serde-common-2.1.3.Final.jar:?]
	at io.apicurio.registry.serde.SchemaResolverConfigurer.configure(SchemaResolverConfigurer.java:75) ~[apicurio-registry-serde-common-2.1.3.Final.jar:?]
	at io.apicurio.registry.serde.AbstractKafkaSerDe.configure(AbstractKafkaSerDe.java:68) ~[apicurio-registry-serde-common-2.1.3.Final.jar:?]
	at io.apicurio.registry.serde.AbstractKafkaDeserializer.configure(AbstractKafkaDeserializer.java:62) ~[apicurio-registry-serde-common-2.1.3.Final.jar:?]
	at io.apicurio.registry.serde.avro.AvroKafkaDeserializer.configure(AvroKafkaDeserializer.java:68) ~[apicurio-registry-serdes-avro-serde-2.1.3.Final.jar:?]
	at org.springframework.kafka.support.serializer.ErrorHandlingDeserializer.configure(ErrorHandlingDeserializer.java:134) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:714) ~[kafka-clients-2.7.0.jar:?]
	... 89 more
Caused by: java.util.NoSuchElementException
	at java.lang.CompoundEnumeration.nextElement(ClassLoader.java:3046) ~[?:?]
	at java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1206) ~[?:?]
	at java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1221) ~[?:?]
	at java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1265) ~[?:?]
	at java.util.ServiceLoader$2.next(ServiceLoader.java:1306) ~[?:?]
	at java.util.ServiceLoader$2.next(ServiceLoader.java:1297) ~[?:?]
	at java.util.ServiceLoader$3.next(ServiceLoader.java:1395) ~[?:?]
	at io.apicurio.registry.rest.client.RegistryClientFactory.resolveProviderInstance(RegistryClientFactory.java:105) ~[apicurio-registry-client-2.1.3.Final.jar:?]
	at io.apicurio.registry.rest.client.RegistryClientFactory.create(RegistryClientFactory.java:77) ~[apicurio-registry-client-2.1.3.Final.jar:?]
	at io.apicurio.registry.rest.client.RegistryClientFactory.create(RegistryClientFactory.java:71) ~[apicurio-registry-client-2.1.3.Final.jar:?]
	at io.apicurio.registry.serde.AbstractSchemaResolver.configure(AbstractSchemaResolver.java:82) ~[apicurio-registry-serde-common-2.1.3.Final.jar:?]
	at io.apicurio.registry.serde.DefaultSchemaResolver.configure(DefaultSchemaResolver.java:56) ~[apicurio-registry-serde-common-2.1.3.Final.jar:?]
	at io.apicurio.registry.serde.SchemaResolverConfigurer.configure(SchemaResolverConfigurer.java:75) ~[apicurio-registry-serde-common-2.1.3.Final.jar:?]
	at io.apicurio.registry.serde.AbstractKafkaSerDe.configure(AbstractKafkaSerDe.java:68) ~[apicurio-registry-serde-common-2.1.3.Final.jar:?]
	at io.apicurio.registry.serde.AbstractKafkaDeserializer.configure(AbstractKafkaDeserializer.java:62) ~[apicurio-registry-serde-common-2.1.3.Final.jar:?]
	at io.apicurio.registry.serde.avro.AvroKafkaDeserializer.configure(AvroKafkaDeserializer.java:68) ~[apicurio-registry-serdes-avro-serde-2.1.3.Final.jar:?]
	at org.springframework.kafka.support.serializer.ErrorHandlingDeserializer.configure(ErrorHandlingDeserializer.java:134) ~[spring-kafka-2.7.9.jar:2.7.9]
	at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:714) ~[kafka-clients-2.7.0.jar:?]
	... 89 more
```

**Comment by carlesarnal** [comment]
I recommend you to set the provider using the RegistryClientFactory#setProvider, which will avoid concurrent calls to the service loader (I assume you already know which implementation of the httpclient you want to use).

**Comment by MosheElisha** [comment]
Thanks @carlesarnal . When using Spring @KafkaListener annotation on multiple beans, it is difficult to control the order and things might run in parallel.
We can invoke RegistryClientFactory#setProvider but we need to know the parameters to pass. When this is done by @KafkaListener, it uses the Kafka properties automatically.

Can you please consider solving this in RegistryClientFactory? It might be as simple as adding `synchronized` on `resolveProviderInstance()`. If it is not meant to run in parallel anyway, the performance penalty is almost zero.

**Comment by carlesarnal** [comment]
You just need to make sure it gets called before your actual producers and consumers are created. I'm not familiar with how the KafkaListener annotation works. You need to pass that method an instance of the provider https://github.com/Apicurio/apicurio-common-rest-client/blob/main/rest-client-jdk/src/main/java/io/apicurio/rest/client/JdkHttpClientProvider.java

**Comment by MosheElisha** [comment]
Thanks, @carlesarnal . 

Spring KafkaListener is very popular. You can see usage example here - https://docs.spring.io/spring-kafka/reference/html/#with-java-configuration-no-spring-boot and you can see how it creates the deserializers in the stacktrace above.

Invoking RegistryClientFactory#setProvider with JdkHttpClientProvider means that I need to force some load order on my Spring beans and I need to pass the Apicurio details (e.g., endpoint) instead of having everything in one place (i.e., ConsumerFactory properties).

Addressing this issue in apicurio-registry will make it easier for everyone using Spring KafkaListener to integrate with Apicurio. 
Why do you object addressing the issue in apicurio-registry?

**Comment by carlesarnal** [comment]
I’m no objecting to address the issue in Registry. What I’m proposing is just a workaround until we actually fix and release it. Is a workaround that should work for existing releases. About the second part of your comment, that’s not 100% accurate. Calling the `setProvider` will only force the factory to use the given provider instead of using the actual service loader mechanism, you don’t need to pass any details like e.g. registry url in that call.

**Comment by MosheElisha** [comment]
Thanks, @carlesarnal ! Sorry, I did not understand that you are suggesting a temporary workaround. Thank you for that.


> About the second part of your comment, that’s not 100% accurate....

OK. Got it. Thanks.

**Comment by carlesarnal** [comment]
@MosheElisha we did a release of registry that also includes some refactoring of how this works. Do you mind giving it a try?

Thanks!

**Comment by MosheElisha** [comment]
Thanks, @carlesarnal ! I have updated to 2.2.0.Final and so far so good.

**Comment by MosheElisha** [comment]
Hi @carlesarnal ,

After I removed the suggested workaround, seems there is still an issue. Seems the issue is in apicurio-common-rest-client-common-0.1.7.Final.jar

```
=> org.apache.kafka.common.KafkaException: Failed to construct kafka producer
   org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:441)
   org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:291)
   org.springframework.kafka.core.DefaultKafkaProducerFactory.createRawProducer(DefaultKafkaProducerFactory.java:801)
   org.springframework.kafka.core.DefaultKafkaProducerFactory.createKafkaProducer(DefaultKafkaProducerFactory.java:642)
   org.springframework.kafka.core.DefaultKafkaProducerFactory.doCreateProducer(DefaultKafkaProducerFactory.java:602)
   [...]
 Caused by: java.lang.IllegalStateException: java.util.NoSuchElementException
   io.apicurio.registry.resolver.AbstractSchemaResolver.configure(AbstractSchemaResolver.java:87)
   io.apicurio.registry.resolver.DefaultSchemaResolver.configure(DefaultSchemaResolver.java:55)
   io.apicurio.registry.serde.SchemaResolverConfigurer.configure(SchemaResolverConfigurer.java:87)
   io.apicurio.registry.serde.AbstractKafkaSerDe.configure(AbstractKafkaSerDe.java:70)
   io.apicurio.registry.serde.avro.AvroKafkaSerializer.configure(AvroKafkaSerializer.java:91)
   [...]
 Caused by: java.util.NoSuchElementException
   java.base/java.util.ServiceLoader$2.next(ServiceLoader.java:1309)
   java.base/java.util.ServiceLoader$2.next(ServiceLoader.java:1297)
   java.base/java.util.ServiceLoader$3.next(ServiceLoader.java:1395)
   io.apicurio.rest.client.spi.ApicurioHttpClientFactory.resolveProviderInstance(ApicurioHttpClientFactory.java:34)
   io.apicurio.rest.client.spi.ApicurioHttpClientFactory.create(ApicurioHttpClientFactory.java:22)
   [...]
```

**Comment by carlesarnal** [comment]
> Hi @carlesarnal ,
> 
> After I removed the suggested workaround, seems there is still an issue. Seems the issue is in apicurio-common-rest-client-common-0.1.7.Final.jar
> 
> ```
> => org.apache.kafka.common.KafkaException: Failed to construct kafka producer
>    org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:441)
>    org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:291)
>    org.springframework.kafka.core.DefaultKafkaProducerFactory.createRawProducer(DefaultKafkaProducerFactory.java:801)
>    org.springframework.kafka.core.DefaultKafkaProducerFactory.createKafkaProducer(DefaultKafkaProducerFactory.java:642)
>    org.springframework.kafka.core.DefaultKafkaProducerFactory.doCreateProducer(DefaultKafkaProducerFactory.java:602)
>    [...]
>  Caused by: java.lang.IllegalStateException: java.util.NoSuchElementException
>    io.apicurio.registry.resolver.AbstractSchemaResolver.configure(AbstractSchemaResolver.java:87)
>    io.apicurio.registry.resolver.DefaultSchemaResolver.configure(DefaultSchemaResolver.java:55)
>    io.apicurio.registry.serde.SchemaResolverConfigurer.configure(SchemaResolverConfigurer.java:87)
>    io.apicurio.registry.serde.AbstractKafkaSerDe.configure(AbstractKafkaSerDe.java:70)
>    io.apicurio.registry.serde.avro.AvroKafkaSerializer.configure(AvroKafkaSerializer.java:91)
>    [...]
>  Caused by: java.util.NoSuchElementException
>    java.base/java.util.ServiceLoader$2.next(ServiceLoader.java:1309)
>    java.base/java.util.ServiceLoader$2.next(ServiceLoader.java:1297)
>    java.base/java.util.ServiceLoader$3.next(ServiceLoader.java:1395)
>    io.apicurio.rest.client.spi.ApicurioHttpClientFactory.resolveProviderInstance(ApicurioHttpClientFactory.java:34)
>    io.apicurio.rest.client.spi.ApicurioHttpClientFactory.create(ApicurioHttpClientFactory.java:22)
>    [...]
> ```

Hi, is that error and intermittent one like it was before or it's something "permanent" each time you try to create a serializer/deserializer?

**Comment by MosheElisha** [comment]
It is sporadic. It probably depends on the exact timing of the parallel threads.

**Comment by EricWittmann** [comment]
@carlesarnal bump :)

**Comment by MosheElisha** [comment]
Thanks, @EricWittmann . 

I encountered the same issue when using Beam and Flink (beam-runners-flink-1.13-job-server-2.36.0.jar) configured with parallelism. 

```
org.apache.beam.sdk.util.UserCodeException: org.apache.kafka.common.KafkaException: Failed to construct kafka producer
...
Caused by: org.apache.kafka.common.KafkaException: Failed to construct kafka producer
...
Caused by: java.lang.IllegalStateException: java.util.NoSuchElementException
...
Caused by: java.util.NoSuchElementException
```
Full stacktrace: [beam-flink-exception.txt](https://github.com/Apicurio/apicurio-registry/files/8221721/beam-flink-exception.txt)

The fix is now even more important because I still did not find a way to invoke RegistryClientFactory#setProvider before Flink configures the AvroKafkaSerializer.

**Comment by carlesarnal** [comment]
The fixed version of the library is now merged into master, in case you want to give it a try.

**Comment by MosheElisha** [comment]
Thanks, @carlesarnal ! Can you please send a link to the commit / PR?

**Comment by carlesarnal** [comment]
Hi, this is the concrete commit: https://github.com/Apicurio/apicurio-registry/commit/2295b7a0a2735c2c6e3e3605a05bff9e2133f267. If you want to test it using docker, you can use `latest-snapshot`.

**Comment by MosheElisha** [comment]
@carlesarnal Thanks a lot!

**Comment by carlesarnal** [comment]
Closing this one as it has been fixed. If it happen again, please, re-open as needed.
